### PR TITLE
Preserve cause of exception when getting/setting attribute in DeviceProxy

### LINKED ISSUE #364: DeviceProxy AttributeError: cause is lost
When a device is not running, trying to get an attribute raises `AttributeError` while losing the cause of this error. Why not use `raise AttributeError() from cause`?

```python
# tango.device_proxy
__DeviceProxy__getattr(self, name):
    cause = None
    ...
    try:
        self.__refresh_cmd_cache()
    except Exception as e:
        cause = e
    ...
    raise AttributeError(name) from cause
```

### PR DESCRIPTION
**Pull request by woutdenolf**
Closes #364

### CODE REVIEW

**Comment by woutdenolf** [comment]
I guess we still need to support python 2?

```python
from six import raise_from
```

**Comment by ajoubertza** [comment]
> I guess we still need to support python 2?

Yes, please!

File: tango/device_proxy.py [hunk 1]
```diff
@@ -293,38 +294,42 @@ def __DeviceProxy__getattr(self, name):
 
     try:
         self.__refresh_cmd_cache()
-    except:
-        pass
+    except Exception as e:
+        if cause is None:
+            cause = e
```
Comment (ajoubertza) [inline tango/device_proxy.py:299]: There would be multiple exceptions in this method if the device is off/unreachable, so keeping the first exception is a good approach 👍

File: tango/device_proxy.py [hunk 2]
```diff
@@ -334,29 +339,36 @@ def __DeviceProxy__setattr(self, name, value):
...
 
     if name_l in self.__get_pipe_cache():
         return self.write_pipe(name, value)
 
-    return super(DeviceProxy, self).__setattr__(name, value)
+    try:
+        return super(DeviceProxy, self).__setattr__(name, value)
+    except Exception as e:
+        raise e from cause
```
Comment (ajoubertza) [inline tango/device_proxy.py]: This case is interesting.  If we get here then `name` isn't a known Tango device command/attribute/pipe.  We end up writing to either an existing Python class instance attribute, or adding a new one, if it doesn't exist yet.

I'm not sure that `__setattr__` can even raise an exception?  If it did, the `cause` may or may not be relevant.

A couple of cases could occur for the example of an unreachable device:
1. We haven't yet cached the lists of commands, attributes and pipes.  
2. We have cached the lists of commands, attributes and pipes prior to this call. 

Then we get these four scenarios, depending on if the Tango command/attribute/pipe either:

- Exists on the device:
  - For case 1:  we'll write to a local class instance attribute and the user won't even get an error.  This behaviour is unexpected, but not a new problem.
  - For case 2: we would have failed because it is a command, or we tried to call `__set_attribute_value` or `write_pipe` earlier in this method, so the user will get whatever error those methods throw.  That's fine.
- Doesn't exist on the device:
  - For case 1:  we'll write to a local class instance attribute and the user won't get an error.  This could be the desired behaviour, if the user was storing some other info on an instance of a `DeviceProxy`.  Or, it could be undesired, if the user thought they were setting a Tango attribute/pipe but had the wrong name.  Not much we can do about that. 
  - For case 2: same as case 1.  There won't be an error, and we're not sure if the user expected one of not.

In conclusion, the change in the PR is fine.  It has just made us aware of some nuances in how setting attributes on a `DeviceProxy` behaves.

File: tango/device_proxy.py [hunk 3]
```diff
@@ -334,29 +339,36 @@ def __DeviceProxy__setattr(self, name, value):
...
+
```
Comment (ajoubertza) [inline tango/device_proxy.py]: We don't need this blank line.  Just 2 between functions according to PEP8.  I'm not sure why Codacy didn't complain.

**Review by ajoubertza — CHANGES_REQUESTED** [review]
The approach looks good, just need the Python 2 support.

File: tango/device_proxy.py [hunk 4]
```diff
@@ -334,29 +339,35 @@ def __DeviceProxy__setattr(self, name, value):
...
     try:
         self.__refresh_cmd_cache()
-    except:
-        pass
+    except Exception as e:
+        if cause is None:
+            cause = e
 
     if name_l in self.__get_cmd_cache():
         raise TypeError('Cannot set the value of a command')
```
Comment (ajoubertza) [inline tango/device_proxy.py]: Shouldn't this line include `cause`, similar to line 332?

**Review by ajoubertza — COMMENTED** [review]
Thanks for the update.  I have one query.

(And I updated the test requirements due to the last CI failure)

File: tango/device_proxy.py (hunk 4 above)
Comment (woutdenolf) [reply tango/device_proxy.py]: Good catch. Thanks!

**Review by ajoubertza — APPROVED** [review]
Looks good.  Thanks for the contribution!
